### PR TITLE
Set MAC state for devices with active session

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

References https://github.com/TheThingsIndustries/lorawan-stack/issues/2417

#### Changes

- For ~OTAA~ devices with a session, set ~current parameters instead of~ MAC state. ~This is only relevant for the Rx1 Delay for devices exported from TTNv2~

#### Testing

Test with devices from TTN

OTAA device, without a session -> Sets mac_settings.rx1_delay
OTAA device with session -> Sets mac_state.current_parameters.rx1_delay
ABP devices -> Sets mac_settings.rx1_delay

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.

### CODE REVIEW

File: pkg/source/ttnv2/source.go [hunk 1]
```diff
@@ -169,6 +168,21 @@ func (s *Source) ExportDevice(devID string) (*ttnpb.EndDevice, error) {
...
+			LoRaWANVersion: ttnpb.MAC_V1_0_2,
+			CurrentParameters: ttnpb.MACParameters{
+				Rx1Delay: ttnpb.RX_DELAY_1,
+			},
+		}
+	} else {
+		v3dev.MACSettings.Rx1Delay = &ttnpb.RxDelayValue{
+			Value: ttnpb.RX_DELAY_1,
+		}
+	}
```
Comment (rvolosatovs) [inline pkg/source/ttnv2/source.go:184]:
```suggestion
	if deviceHasSession {
		v3dev.MACState = &ttnpb.MACState{
			DeviceClass:    ttnpb.CLASS_A,
			LoRaWANVersion: ttnpb.MAC_V1_0_2,
			CurrentParameters: ttnpb.MACParameters{
				Rx1Delay: ttnpb.RX_DELAY_1,
			},
		}
	} 
	if !deviceSupportsJoin {
		v3dev.MACSettings.Rx1Delay = &ttnpb.RxDelayValue{
			Value: ttnpb.RX_DELAY_1,
		}
	}
```

**Review by rvolosatovs — CHANGES_REQUESTED** [review]

**Comment by rvolosatovs** [comment]
If `session` is set, then a complete `mac-state` should be set as well, Network Server should not allow setting just `session`, for example, and not setting `mac-state` as well.

**Review by rvolosatovs — APPROVED** [review]
This looks good, but the tool should be careful with the field paths sent to NS - if the tool sends top-level `mac_state`, then the created device will end up with a corrupted MAC state and will be unusable

**Comment by rvolosatovs** [comment]
And, to reiterate, I think the cleanest approach would be:
1. Generate new MAC state
2. Apply all modifications to it according to v2 specifics
3. Send the whole `mac_state` with top-level `mac_state` field path specified

**Review by htdvisser — APPROVED** [review]
LGTM

File: pkg/source/ttnv2/config.go [hunk 2]
```diff
@@ -123,6 +128,11 @@ func getConfig(ctx context.Context, flags *pflag.FlagSet) (config, error) {
 	}
 	ttnlog.Set(logger)
 
+	fpFetcher, err := fetch.FromHTTP(stringFlag("frequency-plans-url"), true)
```
Comment (rvolosatovs) [inline pkg/source/ttnv2/config.go:131]: Is this flag set?

**Review by rvolosatovs — APPROVED** [review]
A possible improvement is to not set `mac-state.desired-parameters`, such that v3 NS would compute that by itself and modify the state accordingly, but I'm not sure we should do that.
For example, that would mean that immediately after migration v3 NS would try to reconfigure device to use RX1 Delay of 5 seconds(v3 default) via MAC commands, @johanstokking @htdvisser - what do you think?
IMO it's safest not to do that and just let the device stay with exact same configuration as v2 to minimize chance of potential issues. Users can set `mac-state.desired-parameters.rx1-delay` if necessary

**Comment by johanstokking** [comment]
> For example, that would mean that immediately after migration v3 NS would try to reconfigure device to use RX1 Delay of 5 seconds(v3 default) via MAC commands, @johanstokking @htdvisser - what do you think?

Yeah good point.

What are the other settings that we should consider, besides RX1 delay?

As for this particular instance, I do think that we should reconfigure the RX1 delay to TTS' default.

**Comment by neoaggelos** [comment]
Updated according to our discussion with @rvolosatovs. Will proceed with testing after NS supports importing device sessions.

File: pkg/source/ttnv2/source.go [hunk 3]
```diff
@@ -210,5 +209,8 @@ func (s *Source) RangeDevices(appID string, f func(source.Source, string) error)
 
 // Close implements the Source interface.
 func (s *Source) Close() error {
-	return s.client.Close()
+	if s.client != nil {
```
Comment (johanstokking) [inline pkg/source/ttnv2/source.go:212]: I don't really like the fact that we lazy-initialize `client` with the first application ID that is seen by `Source`. Is there a good reason for that?

**Comment by rvolosatovs** [comment]
Blocked by https://github.com/TheThingsIndustries/lorawan-stack/issues/2417

File: pkg/source/ttnv2/source.go (hunk 3 above)
Comment (neoaggelos) [reply pkg/source/ttnv2/source.go:212]: Lazy initialization is removed as part of #13

**Comment by rvolosatovs** [comment]
Blocking https://github.com/TheThingsIndustries/lorawan-stack/issues/2417

File: pkg/source/ttnv2/source.go [hunk 4]
```diff
@@ -169,6 +169,24 @@ func (s *Source) ExportDevice(devID string) (*ttnpb.EndDevice, error) {
...
+	if deviceHasSession {
+		var err error
+		if v3dev.MACState, err = mac.NewState(v3dev, s.config.fpStore, ttnpb.MACSettings{}); err != nil {
+			return nil, err
+		}
+		// Ensure MAC state matches v2 configuration.
+		v3dev.MACState.DeviceClass = ttnpb.CLASS_A
+		v3dev.MACState.LoRaWANVersion = ttnpb.MAC_V1_0_2
+		v3dev.MACState.CurrentParameters.Rx1Delay = ttnpb.RX_DELAY_1
+		v3dev.MACState.DesiredParameters = ttnpb.MACParameters{}
```
Comment (rvolosatovs) [inline pkg/source/ttnv2/source.go:182]: You should not need to be doing this. Instead, this should be covered by the field mask.
Only relevant fields in `mac_state.current_parameters` and in top-level `mac_state` should be specified

**Review by rvolosatovs — CHANGES_REQUESTED** [review]

**Comment by johanstokking** [comment]
@neoaggelos if this is ready for review, request them.

**Comment by neoaggelos** [comment]
I am currently testing against the changes in https://github.com/TheThingsNetwork/lorawan-stack/pull/3680, will request review when ready.

**Review by rvolosatovs — APPROVED** [review]
I don't see the fieldmask in the diff and don't know where to find it, but I assume that is correct

**Comment by rvolosatovs** [comment]
FYI, https://github.com/TheThingsNetwork/lorawan-stack/pull/3680 is still not merged.

**Pull request merged**